73-offline-attestation.md

```diff
@@ -0,0 +1,302 @@
+perform a full attestation, in a persistent external time-series datastore.
+This should also include some proof that a given AIK created on a TPM by an
+`agent` was indeed tied to a given EK, a process that is done by the
+`registrar` and whose responsibility is to store it on a tamper-resistant
+metadatastore (e.g. transparency log)
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this enhancement.  Describe why the change is important and the benefits to users.
+-->
+
+The main motivation for adding this functionality is to give auditors and other
+compliance officers the ability to answer, with a proper degree of certainty
+and trust the following question: did node N had its software stack fully
+attested at date T?  Being date "T" a point time that could be well in the
+past, we cannot rely on the accessibility (or even the existence) of the given
+node. Furthermore we cannot even rely on the accessibility (or even the
+existence) of the server-side components of the Keylime cluster (i.e.,
+`registrar` and `verifier`) and thus need to design with these boundary
+conditions in mind.  
+
+### Goals
+
+<!--
+List the specific goals of the enhancement.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- Add functionality on the `registrar` to record (in a tamper-resistant
+  transparency log) the association between the EK and AIK (i.e.
+`tpm2_makecredential`)
+- Add functionality on the `verifier` to record (in a time-series persistent
+  datastore) all the information needed to perform attestation standlone (i.e.,
+quotes and MB/IMA logs)
+- Add a new CLI which will interface with the aforementioned persistent stores,
+  and will call the main, umodified `verifier` code in order to do post-facto
+attestation.  
+
+### Non-Goals
+
+<!--
+What is out of scope for this enhancement?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- The interaction between the time-series persistent datastore and
+  tamper-resistant transparency log will be done by keylime user/operator.
+Inside the core Keylime, a "plugin" architecture will be adopted (very much
+like the "policies" for Measured Boot) and the implementation details of the
+code which will interact with such stores are outside of the scope.
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+- The `registrar` will be modified to, upon initial `agent` registration -
+  which includes the execution of `tpm2_makecredential` - record the EK, AIK
+into a JSON file, sign it (using the private key generated as part of the
+certificates for mTLS interaction with both `tenant` and `registrar`) and then
```

Comment:
Would it not be better to use a tpm resident key to sign this which in turn can be chained back to the EK / tpm itself? I am not sure what value there is with having non-tpm resident private key sign things attestations.

Reply:
I strongly favor a change in the **core** `keylime` in order to use TPM resident keys instead. However, for the moment, I advocate we separate this change from the whole "offline attestation" enhancement, and simply revisit it, with the appropriate modifications if (when?) we decide on such fundamental change.

Reply:
> I strongly favor a change in the core keylime in order to use TPM resident keys instead.

I think for the TLS connections it does not work, because on how the SSL library in Python works. Ideally we would reverse proxy the registrar and verifier through nginx, haproxy or similar for TLS instead of using the limited Python implementation.

For the used signing key I think it should be a separate one.

Reply:
This is the reason why I would like to have a separate discussion on which kinds of signing keys we should use on `Keylime`

Reply:
I am not following the logic here. Why even bother signing the attestations in the first place, especially with a file resident key. What guarantees does this get us, is this to somehow map the machine measured to the registrar / verifier who requested the attestation?